src/shared/port-provider.js

```diff
@@ -0,0 +1,269 @@
+import { isMessageEqual, SOURCE as source } from './port-util';
+
+/**
+ * @typedef {import('../types/config').SidebarConfig} SidebarConfig
+ * @typedef {import('../types/annotator').Destroyable} Destroyable
+ * @typedef {import('./port-util').Message} Message
+ * @typedef {Message['channel']} Channel
+ * @typedef {Message['port']} Port
+ */
+
+/**
+ * PortProvider creates a `MessageChannel` for the communication between two
+ * frames.
+ *
+ * There are 4 types of frames:
+ * - `host`:  frame where the hypothesis client is initially loaded.
+ * - `guest/s`: frame/s with annotatable content. In some instances the guest
+ *    frame can be the host, in others it is an iframe from where either (1)
+ *    the hypothesis client has been injected or (2) the hypothesis client has
+ *    been configured to act exclusively as a guest (not showing the sidebar).
+ * - `notebook`: it is an another hypothesis app that runs on a separate iframe.
+ * - `sidebar`: the main hypothesis app. It runs on an iframe on a different
+ *    origin than the host and is responsible for the communication with the
+ *    backend (fetching and saving annotations).
+ *
+ * This layout represents the current arrangement of frames:
+ *
+ * `host` frame
+ * |-> `sidebar` iframe
+ * |-> `notebook` iframe
+ * |-> [`guest` iframe/s]
+ *
+ * Currently, we support the communication between the following pair of frames:
+ * - `guest-host`
+ * - `guest-sidebar`
+ * - `host-sidebar`
+ * - `notebook-sidebar`
+ *
+ * `PortProvider` is used only on the `host` frame. The rest of the frames use the
+ * companion class, `PortFinder`. `PortProvider` creates a `MessageChannel`
+ * for two frames to communicate with each other. It also listens to requests for
+ * particular `MessagePort` and dispatches the corresponding `MessagePort`.
+ *
+ *
+ *                 PortProvider                                       |                PortFinder
+ * -------------------------------------------------------------------|-------------------------------------------------------------
+ *
+ * 2. listens to requests of `MessagePort` <---------------------------- 1. request a `MessagePort` using `window.parent.postMessage`
+ *                 |
+ *                 V
+ * 3. sends offers of `MessagePort` using `event.source.postMessage` ---> 4. listens to offers of `MessagePort`
+ *                 |
+ *                 V
+ * 5. send reciprocal port to the `sidebar` frame using the
+ *    `host-sidebar(MessageChannel).sidebar(MessagePort)`
+ *
+ *
+ * In some situations, because `guest` iframe/s load in parallel to the `host`
+ * frame, we can not assume that the code in the `host` frame is executed before
+ * the code in a `guest` frame. Hence, we can't assume that `PortProvider` (in
+ * the `host` frame) is initialized before `PortFinder` (in the `guest` frame).
+ * Therefore, for the `PortFinder`, we implement a polling strategy (sending a
+ * message every N milliseconds) until a response is received.
```

Comment:
The fact that polling is used could be an implementation detail of `PortFinder`, except for the fact that `PortProvider` needs to allow for potentially receiving duplicate requests for a guest port.

Reply:
Are you suggesting to remove this comment from the documentation?